feat(coral): improve browse topic a11y

programmiri opened the pull request:
# About this change - What it does

- commit [Improve pagination.](https://github.com/aiven/klaw/pull/312/commits/35028c0a60f8af4ebb7dceb2d5e34a3890bba28f)
Improves usability for screenreader for pagination. Removes unneccessary handling of tabindex. Adds better tests. 

- commit [Fix main nav list.](https://github.com/aiven/klaw/pull/312/commits/ee6d2000fc9dcc0ab198000a1b2997cd245e7932)
semantic error, forgot to wrap one submenu button in a <li> item

- commit [Introduce custom class for a11y improvments for interactive indicators.](https://github.com/aiven/klaw/pull/312/commits/d5bfad7a92d90d0d5b816437028ce2b2abfae8f4)
- adds a global `accessiblity.module.css` where we add styles etc. to improve accessiblity
- some additional styles still need to be set on component level (see main navigation submenu) but that's only necessary as long as these components are not from the DS
- uses the accessibility classes / css vars

-  commit [Improve a11y for label color contrast](https://github.com/aiven/klaw/pull/312/commits/da42b16f3398a015419d6acc009c4ba55dac9e76)
-> darker color for labels to match contrast requirements. reduced the font-size to still be distinguishable from other text

- [Improve a11y for login redirect.](https://github.com/aiven/klaw/pull/312/commits/a15a6464f31c0bf6aba4d9dfe80f948d06055447) 

This adds a) a bit of style to the redirect screen and b) accessibility


## This improves:
- accessibility for visual impaired users (that's also e.g. users in a work environment that reduces contrast on screens)
- accessibility for keyboard users, they now can always see where they are navigating
- accessibility for screenreaders (pagination)

## Not solved yet to go not too far from DS
- buttons (primary and in pagination) are not matching color requirements (not enough contrast). 


## Recordings

### Keyboard navigation
(tabbing through to show focus states)


https://user-images.githubusercontent.com/943800/207083741-bbeef9d8-c00d-493e-a02e-0e441eb23b9d.mov

### Screenreader using pagination 
( ⚠️ sound! best "watch" without actually looking at video, matches user experience better)


https://user-images.githubusercontent.com/943800/207084030-f715bf4d-28b8-44dd-b5ee-093a46dfa9ff.mov


### alert for redirect

Videos are having sound & best watched without watching :D so only listening! 

### Without a11y

https://user-images.githubusercontent.com/943800/207275524-565d57f3-2db5-4920-aafe-cff6607bac08.mov


### With a11y 


https://user-images.githubusercontent.com/943800/207275571-d0b95366-913a-4a82-928c-e05b676a6045.mov


I changed the text because the redirect screen is not there too long and dependent on speed of the screenreader, a user may not get all text read, so I started with the most important info.


## Comit message
title: `feat(coral): improve browse topic a11y`
body: 
```
- Improve usability for screenreader for pagination.
   Removes unneccessary handling of tabindex. Adds better tests.
- Fix main nav list semantics
- Introduce and use custom class for a11y improvments 
   for interactive indicators.
- Improve a11y for login redirect by using alertdialog
```

mathieu-anderson commented on coral/src/services/api-mocks/browser.ts:
```diff
@@ -1,5 +1,9 @@
-const worker = setupWorker();
+// eslint-disable-next-line
+let worker: any;
+if (typeof jest === "undefined") {
```
Candid question because I don't understand: what was the reason for this addition? In which instance is `jest` `undefined` and why do we call `setupWorker` only when it is the case?

programmiri replied:
Good question! Probably should have added a comment. 

This is a little hack to use the mocked API to show pagination in the app, without effecting tests. 
jest is `undefined` in the browser. Without that condition, `setupWorker` will be called when the code runs in tests and that will throw an error. This will not be committed in the end!

mathieu-anderson commented on coral/src/app/main.module.css:
```diff
@@ -1,17 +1,3 @@
-  font-size: 18px;
-}
-
 /* @TODO check design system */
```
Should we also have a constant in `accessibility` for the color in the `body` rule underneath?

```
body {
  margin: 0;
  /*@TODO check with design, it is supposed*/
  /*  to be grey-60 but that does not meet*/
  /*  accessibility standards*/
  /*represents grey-70*/
  color: #4a4b57;
  -webkit-text-size-adjust: 100%;
}
```

mathieu-anderson reviewed (commented):
All the changes in this PR seem good to me, I don't see a reason to leave some of them out, especially with the addition of the `accessibiilty` styles. I left a few questions ^^

programmiri commented on coral/src/app/features/browse-topics/BrowseTopics.test.tsx:
```diff
@@ -189,7 +189,7 @@ describe("BrowseTopics.tsx", () => {
       await waitForElementToBeRemoved(screen.getByText("Loading..."));
       const pagination = screen.queryByRole("navigation", {
-        name: "Pagination",
+        name: /Pagination/,
```
Oh, very good question! 

In this case, I wanted to make sure that no `Pagination` is rendered at all. The paginations accessible name shows the current page and total page when it's rendered. I could test for `"Pagination navigation, you're on page 1 of 1."` here, to be more precise. 

But if the `aria-label` in pagination would change at some point, e.g. be `"Pagination navigation, you're on page 1 of 1"` AND for some reason the pagination would be rendered for a one-page response, this test would still be green. I wanted to make sure that it fails also in that case, to make it a bit more stable. 

Usually, I use the exact string. Exceptions can be an example like this (where I think using the exact string makes the test less safe) or when the exact string does not really matter too much. 

(And sometimes I may at least think about using a regex when the component I test is not directly responsible for setting that string, just to reduce maintenance work, but I don't know if there's an example in the tests right now for that 🤔 )

programmiri commented:
Updated the PR!

programmiri commented:
Added another commit: 

- [Improve a11y for login redirect.](https://github.com/aiven/klaw/pull/312/commits/a15a6464f31c0bf6aba4d9dfe80f948d06055447) 

This adds a) a bit of style to the redirect screen and b) accessibility

Here's a screenrecording of the current style (cave: I changed text after the recording) without and with a11y improvements. 

Videos are having sound & best watched without watching :D so only listening! 

### Without a11y

https://user-images.githubusercontent.com/943800/207275524-565d57f3-2db5-4920-aafe-cff6607bac08.mov


### With a11y 


https://user-images.githubusercontent.com/943800/207275571-d0b95366-913a-4a82-928c-e05b676a6045.mov



I changed the text because the redirect screen is not there too long and dependent on speed of the screenreader, a user may not get all text read, so I started with the most important info.

mathieu-anderson commented on coral/src/app/components/AuthenticationRequiredBoundary.tsx:
```diff
@@ -21,7 +22,31 @@ class AuthenticationRequiredBoundary extends Component<{
+            aria-describedby={"authentication-required-text"}
+          >
+            <Typography.Heading color={"secondary-100"}>
+              <span id={"authentication-required-heading"}>
```
What is the purpose of the `id` attribute? is it related to accessibiility?

programmiri replied:
Yes! The div with `role="alertdialog"` has an `aria-labelledby` and an `aria-describedby` attribute. There I use the `id`s from the elements that contain the accessible label and description. That way, e.g. a screenreader know what to read out here related to the `alertdialog`.

mathieu-anderson reviewed (approved):
I like the new look of the redirect page, on top of the accessibility improvements ^^ Just left one question, but I'm approving this, even if it's not a proper PR.

mathieu-anderson reviewed (approved):
Excellent! Will merge.
